samza-test/src/test/java/org/apache/samza/test/processor/TestZkLocalApplicationRunner.java

```diff
@@ -0,0 +1,498 @@
+  }
+
+  private static class TestKafkaEvent implements Serializable {
+
+    // Actual content of the event.
+    private String eventData;
+
+    // Contains Integer value, which is greater than previous message id.
+    private String eventId;
+
+    TestKafkaEvent(String eventId, String eventData) {
+      this.eventData = eventData;
+      this.eventId = eventId;
+    }
+
+    String getEventId() {
+      return eventId;
+    }
+
+    String getEventData() {
+      return eventData;
+    }
+
+    @Override
+    public String toString() {
+      return eventId + "|" + eventData ;
+    }
+
+    static TestKafkaEvent fromSerializedString(String message) {
+      String[] messageComponents = message.split("|");
+      return new TestKafkaEvent(messageComponents[0], messageComponents[1]);
+    }
+  }
+
+  private static class TestStreamApplication implements StreamApplication {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(TestStreamApplication.class);
+
+    private final String inputTopic;
+
+    private final String outputTopic;
+
+    private final CountDownLatch processedMessagesLatch;
+
+    private final AppLifeCycleListener appLifeCycleListener;
+
+    TestStreamApplication(String inputTopic, String outputTopic, String processorId,
+        CountDownLatch processedMessagesLatch, AppLifeCycleListener appLifeCycleListener) {
+      this.inputTopic = inputTopic;
+      this.outputTopic = outputTopic;
+      this.processedMessagesLatch = processedMessagesLatch;
+      this.appLifeCycleListener = appLifeCycleListener;
+    }
+
+    @Override
+    public void init(StreamGraph graph, Config config) {
+      try {
+        BiFunction<String, String, String> messageBuilder = (key, msg) -> {
+          String result = "";
+          try {
+            TestKafkaEvent incomingMessage = TestKafkaEvent.fromSerializedString(msg);
+            if (appLifeCycleListener != null) {
+              appLifeCycleListener.onMessageReceived(incomingMessage);
+            }
+            result = incomingMessage.toString();
+          } catch (Exception e) {
+            LOGGER.error("Exception when processing message in TestStreamApplication: ", e);
+            appLifeCycleListener.onExceptionOccurred(e);
+          }
+          if (processedMessagesLatch != null) {
+            processedMessagesLatch.countDown();
+          }
+          return result;
+        };
+        MessageStream<String> inputStream = graph.getInputStream(inputTopic, messageBuilder);
+        OutputStream<String, String, String> outputStream = graph.getOutputStream(outputTopic, event -> null, event -> event);
+        inputStream.sendTo(outputStream);
+      } catch (Exception e) {
+        if (appLifeCycleListener != null) {
+          appLifeCycleListener.onExceptionOccurred(e);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void setUp() {
+    super.setUp();
+    testStreamAppName = String.format("test-app-name-%s", UUID.randomUUID().toString());
```

Comment:
Why not just generate random ID once and use for everything so that they can be related?